pkg/client/delete.go

```diff
@@ -0,0 +1,107 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	kubeerror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	clusterRoleFieldName  = "component"
+	clusterRoleFieldValue = "sonobuoy"
+
+	e2eNamespacePrefix = "e2e-"
+)
+
+func (c *SonobuoyClient) Delete(cfg *DeleteConfig, client kubernetes.Interface) error {
+	// Delete the namespace
+	log := logrus.WithFields(logrus.Fields{
+		"kind":      "namespace",
+		"namespace": cfg.Namespace,
+	})
+
+	err := client.CoreV1().Namespaces().Delete(cfg.Namespace, &metav1.DeleteOptions{})
+	if err := logDelete(log, err); err != nil {
+		return errors.Wrap(err, "couldn't delete namespace")
+	}
+
+	if cfg.EnableRBAC {
+		// ClusterRole and ClusterRoleBindings aren't namespaced, so delete them seperately
+		selector := metav1.AddLabelToSelector(
+			&metav1.LabelSelector{},
+			clusterRoleFieldName,
+			clusterRoleFieldValue,
+		)
+
+		deleteOpts := &metav1.DeleteOptions{}
+		listOpts := metav1.ListOptions{
+			LabelSelector: metav1.FormatLabelSelector(selector),
+		}
+
+		err := client.RbacV1().ClusterRoleBindings().DeleteCollection(deleteOpts, listOpts)
+		if err := logDelete(logrus.WithField("kind", "clusterrolebindings"), err); err != nil {
+			return errors.Wrap(err, "failed to delete cluster role binding")
+		}
+
+		// ClusterRole and ClusterRole bindings aren't namespaced, so delete them manually
+		err = client.RbacV1().ClusterRoles().DeleteCollection(deleteOpts, listOpts)
+		if err := logDelete(logrus.WithField("kind", "clusterroles"), err); err != nil {
+			return errors.Wrap(err, "failed to delete cluster role")
+		}
+	}
+
+	// Delete any dangling E2E namespaces
+	namespaces, err := client.CoreV1().Namespaces().List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list namespaces")
+	}
+
+	for _, namespace := range namespaces.Items {
+		if strings.HasPrefix(namespace.Name, e2eNamespacePrefix) {
+
+			log := logrus.WithFields(logrus.Fields{
+				"kind":      "namespace",
+				"namespace": namespace.Name,
+			})
+			err := client.CoreV1().Namespaces().Delete(namespace.Name, &metav1.DeleteOptions{})
+			if err := logDelete(log, err); err != nil {
+				return errors.Wrap(err, "couldn't delete namespace")
+			}
+		}
+	}
+	return nil
+}
+
+func logDelete(log logrus.FieldLogger, err error) error {
```

Comment:
i'm not super stoked on passing a logrus.FieldLogger in here...I see why you did it like this, but it's not my favorite.

This function feels like it's doing too much maybe, it's both logging and handling the kube error. Maybe the logging could happen after the delete takes place, something like:

```
if err := checkError(deleteNamespace()); err != nil {
    return err
}
logrus.Info("deleted namespace")
```

and then we don't have to worry about passing the log around? Another argument for this is that I don't think the user cares if it's already deleted vs deleted vs delete in progress.

Reply:
I guess it feels normal to me, because in other projects I've worked on the logger is _always_ the first argument you use. 

What you proposed is not equivalent, because then we'd be missing the `kind=namespace namespace=heptio-test`

Reply:
It would be straightforward to add that extra data to the logrus.Info line as it would all be in scope.

I think we should talk about or logging strategy in sonobuoy. We likely used the standard logger for expedience (I was hoping this project could get away with the standard logger) but maybe it's time to formalize it more and define a logger to use. 

I still think this function is doing both error checking and logging and should be broken up, but it's ok to move forward. I need to get the demo to working today.